sources/pciclient/src/lib.rs

```diff
@@ -0,0 +1,486 @@
+    /// Call `lspci`, it takes in the command line arguments and return the result as vector of strings.
+    /// # Examples
+    ///
+    /// ```
+    /// let pci_client = PciClient{};
+    /// let lspci_output = pci_client.execute(list_devices_param.to_command_args());
+    /// ```
+    fn execute<I, S>(self, args: I) -> Result<Vec<String>>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        let result = Command::new(LSPCI_PATH)
+            .args(args)
+            .output()
+            .context(CommandFailureSnafu)?;
+        ensure!(
+            result.status.success(),
+            ExecutionFailureSnafu {
+                reason: String::from_utf8_lossy(&result.stderr),
+            }
+        );
+
+        let output = String::from_utf8_lossy(&result.stdout)
+            .lines()
+            .map(str::to_string)
+            .collect();
+        Ok(output)
+    }
+}
+
+/// Internal usage, adding command_executor as parameter allows us to better unit test.
+/// For external usage, check [`list_devices`].
+fn call_list_devices<T: CommandExecutor>(
+    command_executor: T,
+    list_devices_param: ListDevicesParam,
+) -> Result<Vec<ListDevicesOutput>> {
+    let list_device_output = command_executor.execute(list_devices_param.to_command_args())?;
+    parse_list_devices_output(list_device_output)
+}
+
+/// Call `lspci` and check if there is any EFA device attached.
+/// Internal usage, adding command_executor as parameter allows us to better unit test.
+/// For external usage, check [`is_efa_attached`].
+fn check_efa_attachment<T: CommandExecutor>(command_executor: T) -> Result<bool> {
+    let list_devices_param = ListDevicesParamBuilder::new()
+        .vendor(AMAZON_VENDOR_CODE)
+        .build();
+    let list_device_results = call_list_devices(command_executor, list_devices_param)?;
+    Ok(list_device_results
+        .iter()
+        .any(|device_info| device_info.device.contains(EFA_KEYWORD)))
+}
+
+/// Query the list of the devices with options to filter the output.
+/// # Input
+/// list_devices_param: [`ListDevicesParam`]. We will under the hood convert the [`ListDevicesParam`]
+/// to the low level arguments to `lspci`.
+/// According to https://man7.org/linux/man-pages/man8/lspci.8.html#MACHINE_READABLE_OUTPUT,
+/// > If you intend to process the output of lspci automatically,
+/// > please use one of the machine-readable output formats (-m, -vm, -vmm)
+/// > described in this section. All other formats are likely to change between versions of lspci.
+///
+/// * We will use `-m` to get the machine-readable format which ensures compatibility.
+/// * We will use `-n` to get the numeric output for the vendor code and device codes.
+/// * We will optionally insert `-d <selection_expression>` to filter the output.
+///
+/// # Output
+/// Output is [`Result<Vec<ListDevicesOutput>>`], the [`ListDevicesOutput`] is based on
+/// the output format that is decorated by "-n -m".
+///
```

Comment:
These docs are helpful for the nitty gritty, but it doesn't tell me really anything about how you expect one to use this function, which is actually the most important part of this docstring.